[1.20 regression] loss of type inference of chained-comparison operands

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

[My repo](https://gitlab.com/TTsangSC/pytest-autoprofile)[^disclaimer-not-a-plug] has a [line of code](https://gitlab.com/TTsangSC/pytest-autoprofile/-/blob/v0.19.1/.mock-ci/pytest_autoprofile_local_ci/commands.py?ref_type=tags#L54) which uses a chained comparison to check if two `str | None` operands are equal and are not `None` (see example below). Said code failed type-checking in [yesterday's pipeline](https://gitlab.com/TTsangSC/pytest-autoprofile/-/jobs/13742751853#L121) which pulled the latest `mypy` from PyPI. Apparently, since 1.20.0[^disclaimer-version-inconsistency] the `a` argument is no longer narrowed to a `str`.

[^disclaimer-not-a-plug]: Provided for context only since the issue template encourages so: `If the project you encountered the issue in is open source, please provide a link to the project.`

[^disclaimer-version-inconsistency]: Note that I failed to replicate the differing behaviors between 1.19 and 1.20 on `mypy` Playground. Instead, "mypy latest (1.20.0)" retained the behavior from 1.19, while "mypy master branch" replicated the narrowing failure I saw with 1.20.0 in the above pipeline and on my local machine.

**To Reproduce**

Gist URL: https://gist.github.com/mypy-play/dc825cee3a47c4b77265efb1e253ec02

```python
from typing import reveal_type


def some_func(a: str | None = None, b: str | None = None) -> None:
    if None is not a == b:
        reveal_type(a)
        reveal_type(b)
        print(a + b)  # Stand-in for some code using them as strings
    else:
        ...  # Some other processing happens here
```

**Expected Behavior**

<!--
How did you expect mypy to behave? It’s fine if you’re not sure your understanding is correct.
Write down what you thought would happen. If you expected no errors, delete this section.
-->

`a` should narrow to `str`, and also `b` by virtue of its equality with `a`.[^disclaimer-string-subtype]

[^disclaimer-string-subtype]: As noted in **Actual Behavior**, `b` is not narrowed in any recent version. Maybe `mypy` is trying to be more conservative in the narrowing, accommodating for the off-chance of some `str` subclass overriding `.__eq__()` to somehow compare to `True` with `None`... ?  May have to do with (the discussion around) #10708.

**Actual Behavior**

<!-- What went wrong? Paste mypy's output. -->

Below are the `mypy` Playground outputs:

*"mypy master branch"*

```log
main.py:6: note: Revealed type is "str | None"
main.py:7: note: Revealed type is "str | None"
main.py:8: error: Unsupported operand types for + ("str" and "None")  [operator]
main.py:8: error: Unsupported left operand type for + ("None")  [operator]
main.py:8: note: Both left and right operands are unions
Found 2 errors in 1 file (checked 1 source file)
```

*"mypy latest (1.20.0)"*

```log
main.py:6: note: Revealed type is "builtins.str"
main.py:7: note: Revealed type is "builtins.str | None"
main.py:8: error: Unsupported operand types for + ("str" and "None")  [operator]
main.py:8: note: Right operand is of type "str | None"
Found 1 error in 1 file (checked 1 source file)
```

Both versions failed to narrow `b`, but `master` didn't infer anything about `a` either. I'm probably out of my depth here, but I'm suspecting that the chained comparison is incorrectly parsed in the new version into something equivalent to `None is not (a == b)`, instead of the correct `(None is not a) and (a == b)`, leading to the complete loss of type narrowing.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

*For the minimal reproducible example*

- Mypy version used: master (on Playground)
- Mypy command-line flags: nil
- Mypy configuration options from `mypy.ini` (and other config files): nil
- Python version used: 3.14

*For the aforementioned failed pipeline*

- Mypy version used: 1.20.0
- Mypy command-line flags: nil
- Mypy configuration options from `mypy.ini` (and other config files): `ignore_missing_imports = true`  plus [misc. options for file in-/ex-clusion](https://gitlab.com/TTsangSC/pytest-autoprofile/-/blob/v0.19.1/pyproject.toml?ref_type=tags#L109)
- Python version used: 3.14.0

<!-- You can freely edit this text, please remove all the lines you believe are unnecessary. -->

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -6558,6 +6558,7 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
         operands = [collapse_walrus(x) for x in node.operands]
         operand_types = []
         narrowable_operand_index_to_hash = {}
+        narrowable_operand_hash_to_index = {}
         for i, expr in enumerate(operands):
             if not self.has_type(expr):
                 return {}, {}
@@ -6582,6 +6583,7 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
                 h = literal_hash(expr)
                 if h is not None:
                     narrowable_operand_index_to_hash[i] = h
+                    narrowable_operand_hash_to_index[h] = i
 
         # Step 2: Group operands chained by either the 'is' or '==' operands
         # together. For all other operands, we keep them in groups of size 2.
@@ -6672,6 +6674,18 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
                 if_map, else_map = else_map, if_map
 
             partial_type_maps.append((if_map, else_map))
+
+            # Chained comparisons are conjunctions evaluated left-to-right. Feed what we learned
+            # from earlier true comparisons into later comparisons, similarly to `and`.
+            if len(simplified_operator_list) > 1:
+                for expr, expr_type in if_map.items():
+                    h = literal_hash(expr)
+                    if h is None or h not in narrowable_operand_hash_to_index:
+                        continue
+                    operand_index = narrowable_operand_hash_to_index[h]
+                    operand_types[operand_index] = meet_types(
+                        operand_types[operand_index], expr_type
+                    )
 
         # If we have found non-trivial restrictions from the regular comparisons,
         # then return soon. Otherwise try to infer restrictions involving `len(x)`.

--- a/test-data/unit/check-narrowing.test
+++ b/test-data/unit/check-narrowing.test
@@ -3264,7 +3264,7 @@ def bad_but_should_pass(has_key: bool, key: bool, s: tuple[bool, ...]) -> None:
         reveal_type(key)  # N: Revealed type is "builtins.bool"
 [builtins fixtures/primitives.pyi]
 
-[case testNarrowChainedComparisonMeet]
+[case testNarrowChainedComparisonMeetAndForwardPropagation]
 # flags: --strict-equality --warn-unreachable
 from __future__ import annotations
 from typing import Any
@@ -3272,7 +3272,7 @@ from typing import Any
 def f1(a: str | None, b: str | None) -> None:
     if None is not a == b:
         reveal_type(a)  # N: Revealed type is "builtins.str"
-        reveal_type(b)  # N: Revealed type is "builtins.str | None"
+        reveal_type(b)  # N: Revealed type is "builtins.str"
 
     if (None is not a) and (a == b):
         reveal_type(a)  # N: Revealed type is "builtins.str"
@@ -3290,11 +3290,33 @@ def f2(a: Any | None, b: str | None) -> None:
 def f3(a: str | None, b: Any | None) -> None:
     if None is not a == b:
         reveal_type(a)  # N: Revealed type is "builtins.str"
-        reveal_type(b)  # N: Revealed type is "Any | builtins.str | None"
+        reveal_type(b)  # N: Revealed type is "Any | builtins.str"
 
     if (None is not a) and (a == b):
         reveal_type(a)  # N: Revealed type is "builtins.str"
         reveal_type(b)  # N: Revealed type is "Any | builtins.str"
+
+def f4(a: str | None, b: str | None, c: str | None) -> None:
+    if None is not a == b == c:
+        reveal_type(a)  # N: Revealed type is "builtins.str"
+        reveal_type(b)  # N: Revealed type is "builtins.str"
+        reveal_type(c)  # N: Revealed type is "builtins.str"
+
+    if (None is not a) and (a == b) and (b == c):
+        reveal_type(a)  # N: Revealed type is "builtins.str"
+        reveal_type(b)  # N: Revealed type is "builtins.str"
+        reveal_type(c)  # N: Revealed type is "builtins.str"
+
+def f5(pair: tuple[None, int] | tuple[str, str], other: str | None) -> None:
+    if None is not pair[0] == other:
+        reveal_type(pair[0])  # N: Revealed type is "builtins.str"
+        reveal_type(pair)  # N: Revealed type is "tuple[builtins.str, builtins.str]"
+        reveal_type(other)  # N: Revealed type is "builtins.str"
+
+    if (None is not pair[0]) and (pair[0] == other):
+        reveal_type(pair[0])  # N: Revealed type is "builtins.str"
+        reveal_type(pair)  # N: Revealed type is "tuple[builtins.str, builtins.str]"
+        reveal_type(other)  # N: Revealed type is "builtins.str"
 [builtins fixtures/primitives.pyi]
 
 [case testNarrowTypeObject]

